tests/models/TemplateConfigurationTest.php

```diff
@@ -22,4 +22,94 @@ public function testCopyMinimalTemplate()
+
+        // Check that invalid paths (real or virtual paths pointing to existing files outside of the allowed folders) are marked as invalid
+        $this->assertEquals('invalid:' . $options['existingInvalidFile'], $savedOptions['existingInvalidFile']);
+        $this->assertEquals('invalid:' . $options['virtualPathWithTraversalOutsideTheme'], $savedOptions['virtualPathWithTraversalOutsideTheme']);
+
+        // Check that non-paths (values that don't match an existing file) are not changed
+        $this->assertEquals($options['nonExistingFile'], $savedOptions['nonExistingFile']);
+    }
+
+    public function testImageSrc()
+    {
+        $templateConfiguration =  \Template::getInstance('fruity', null, null, false);
+
+        $options = [
+            'realThemeFile' => 'themes/survey/fruity/files/logo.png',
+            'realGeneralFile' => 'upload/themes/survey/generalfiles/index.html',
+            'realRelativeThemeFile' => './files/logo.png',
+            'virtualThemeFile' => 'image::theme::files/logo.png',
+            'virtualGeneralFile' => 'image::generalfiles::index.html',
+            'nonExistingFile' => 'test.png',
+            'existingInvalidFile' => 'themes/admin/Apple_Blossom/images/Limesurvey_logo.png',
+            'virtualPathWithTraversalInsideTheme' => 'image::theme::../fruity/files/logo.png',
+            'virtualPathWithTraversalOutsideTheme' => 'image::theme::../vanilla/files/logo.png',
+        ];
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(realThemeFile) }}', $options, $templateConfiguration);
+        $this->assertEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(realGeneralFile) }}', $options, $templateConfiguration);
+        $this->assertEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(realRelativeThemeFile) }}', $options, $templateConfiguration);
+        $this->assertNotEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(virtualThemeFile) }}', $options, $templateConfiguration);
+        $this->assertNotEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(nonExistingFile) }}', $options, $templateConfiguration);
+        $this->assertEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(existingInvalidFile) }}', $options, $templateConfiguration);
+        $this->assertEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(virtualPathWithTraversalInsideTheme) }}', $options, $templateConfiguration);
+        $this->assertNotEmpty($output);
+
+        $output = \Yii::app()->twigRenderer->convertTwigToHtml('{{ imageSrc(virtualPathWithTraversalOutsideTheme) }}', $options, $templateConfiguration);
+        $this->assertEmpty($output);
+
+        // There is no general image file uploaded but, when setting a valid default,
```

Comment:
This is hard to test as general files is empty at time of running the test.
So, we use the "default" parameter of imageSrc.